conduit/data/datasets/base.py

```diff
@@ -21,23 +32,42 @@
 )
 from conduit.logging import init_logger
 
-__all__ = ["CdtDataset"]
+__all__ = [
+    "CdtDataset",
+    "I",
+    "S",
+    "X",
+    "Y",
+]
+
+
+I = TypeVar(
+    "I",
+    NamedSample[LoadedData],
+    BinarySample[LoadedData],
+    SubgroupSample[LoadedData],
+    TernarySample[LoadedData],
+)
+
+X = TypeVar("X", bound=UnloadedData)
+S = TypeVar("S", bound=Optional[Tensor])
+Y = TypeVar("Y", bound=Optional[Tensor])
 
 
-class CdtDataset(DatasetProt[NamedSample]):
+class CdtDataset(DatasetProt[I], Generic[I, X, Y, S]):
     _repr_indent: ClassVar[int] = 4
     _logger: Optional[logging.Logger] = None
 
     def __init__(
-        self, *, x: UnloadedData, y: Optional[TargetData] = None, s: Optional[TargetData] = None
+        self, *, x: X, y: Optional[TargetData] = None, s: Optional[TargetData] = None
     ) -> None:
         self.x = x
         if isinstance(y, np.ndarray):
             y = torch.as_tensor(y)
         if isinstance(s, np.ndarray):
             s = torch.as_tensor(s)
-        self.y = y if y is None else y.squeeze()
-        self.s = s if s is None else s.squeeze()
+        self.y: S = y if y is None else y.squeeze()
```

Comment:
```suggestion
        self.y: Y = y if y is None else y.squeeze()
```

Reply:
Did you address this?

Reply:
I committed your suggestion (or tried to)